lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2041,6 +2041,108 @@ namespace {
+        return;
+      }
+
+      clang::CXXConstructorDecl *defaultCtorDecl = nullptr;
+      for (clang::CXXConstructorDecl *ctor : cxxRecordDecl->ctors()) {
+        if (ctor->isDefaultConstructor() && !ctor->isDeleted()) {
+          defaultCtorDecl = ctor;
+          break;
+        }
+      }
+
+      if (!defaultCtorDecl)
+        return;
+
+      clang::IdentifierTable &idents = clangCtx.Idents;
+      clang::IdentifierInfo *generatedFuncName = &idents.get(
+          ("__returns_" + cxxRecordDecl->getNameAsString()).c_str());
+
+      clang::FunctionProtoType::ExtProtoInfo EPI;
+      clang::QualType generatedFuncTy =
+          clangCtx.getFunctionType(cxxRecordPtrTy, {}, EPI);
+
+      clang::CXXMethodDecl *generatedCxxMethodDecl =
+          clang::CXXMethodDecl::Create(
+              clangCtx, cxxRecordDecl, cxxRecordDecl->getLocation(),
+              clang::DeclarationNameInfo(generatedFuncName,
+                                         cxxRecordDecl->getLocation()),
+              generatedFuncTy, nullptr, clang::SC_Static,
+              /*UsesFPIntrin*/ false, /*isInline=*/true,
+              clang::ConstexprSpecKind::Unspecified,
+              cxxRecordDecl->getLocation());
+
+      generatedCxxMethodDecl->setAccess(clang::AccessSpecifier::AS_public);
+
+      clang::SwiftAttrAttr *generatedReturnsRetainedAttr =
+          clang::SwiftAttrAttr::Create(clangCtx, "returns_retained");
+      generatedCxxMethodDecl->addAttr(generatedReturnsRetainedAttr);
+
+      clang::SwiftNameAttr *generatedSwiftNameInitAttr =
+          clang::SwiftNameAttr::Create(clangCtx, "init()");
+      generatedCxxMethodDecl->addAttr(generatedSwiftNameInitAttr);
+
+      clang::ExprResult generatedConstructExpr =
+          clangSema.BuildCXXConstructExpr(
+              clang::SourceLocation(), cxxRecordTy, defaultCtorDecl,
+              /*Elidable*/ false, clang::MultiExprArg(),
+              /*HadMultipleCandidates,*/ false,
+              /*IsListInitialization*/ false,
+              /*IsStdInitListInitialization*/ false,
+              /*RequiresZeroInit*/ true, // TODO: fix the propagation of this
+                                         // zeroinit flag to BuildCXXNew
+              clang::CXXConstructionKind::Complete, clang::SourceRange());
+
+      assert(!generatedConstructExpr.isInvalid() &&
+             "generatedConstructExpr is isInvalid in "
+             "generateStaticFactoryForCXXForeignRef");
+
+      clang::ExprResult generatedNewExprResult = clangSema.BuildCXXNew(
+          clang::SourceRange(), /*UseGlobal*/ false, clang::SourceLocation(),
+          {}, clang::SourceLocation(), clang::SourceRange(), cxxRecordTy,
+          clangCtx.getTrivialTypeSourceInfo(cxxRecordTy), std::nullopt,
+          clang::SourceRange(), generatedConstructExpr.get());
+
+      assert(!generatedNewExprResult.isInvalid() &&
+             "generatedNewExprResult is isInvalid in "
+             "generateStaticFactoryForCXXForeignRef");
```

Comment:
The assert failure usually points out the location, we do not need to repeat this in the assert string.